mautic/mautic
# True emoji support for notifications

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | N
| Automated tests included? | /
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is migrating out of EmojiHelper and its transliteration and instead implements UTF8MB4 encoding on those database columns. This way marketers can use emojis without fear of database errors.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
0. In order to create new push notifications through the UI you have to enable the OneSignal plugin with any random credentials.
1. Save a push notification with this text in heading and message: `🏢 City Building`
2. You will get a SQL error.

#### Steps to test this PR:
1. Downgrade your encoding on some columns to Mautic 2 encoding to test the migration fully:
```sql
ALTER TABLE push_notifications CHANGE heading heading LONGTEXT CHARACTER SET utf8 COLLATE utf8_unicode_ci;
ALTER TABLE push_notifications CHANGE message message LONGTEXT CHARACTER SET utf8 COLLATE utf8_unicode_ci;
```
2. Run the migration
3. Check the encoding for push_notifications.heading, push_notifications.message have UTF8MB4 encoding.
5. Repeat the tests to reproduce and the push notification should be saved properly.

#### Other areas of Mautic that may be affected by the change:
1. Push Notifications
2. Core bundle notifications

#### List deprecations along with the new alternative:
1. None

#### List backwards compatibility breaks:
1. None
